### PR TITLE
Convert a prompt to a "Zoom Video"

### PR DESCRIPTION
**Pull request by veprogames**
dream_video.py creates a video by running img2img for each movie frame and cropping and rescaling the image (zooming in)

dream_video.py can be called on the CLI like this
`python scripts/dream_video.py "prompt" -F <frame count> [-I <initial-image>]`
and should be useable as a module:
```py
import dream_video

dream_video.prompt2vid(prompt="Prompt", n_frames=120)
```

- [X] Create Frames in outputs/vid-samples
- [x] Individual folders in vid-samples per prompt
- [X] Create a .mp4 with cv2 VideoWriter (works at least on Windows)
- [x] Allow more Parameters (cfg_scale, FPS)
- [x] Adjust for new changes (e. g. ldm.generate)
- [x] More video smoothness (see https://www.reddit.com/r/MediaSynthesis/comments/x2yqlj/poisons_of_civilization_deforum_stable_diffusion/) -> maybe a constant seed works?
- [x] Option to use contant or random seed per frame
- (maybe) Integrate into dream.py CLI

### CODE REVIEW

**Comment by veprogames** [comment]
ok, after trying it out it just seems to turn into an almost single colored image:


shouldn't it change more heavily instead of converging to a single color?

**Comment by veprogames** [comment]
the folder name is used as is, no escaping right now.

**Comment by veprogames** [comment]
I don't know if the video creation is working on Linux yet (if it fails, you can still merge the frames with ffmpeg). Is there error handling missing? There should be default values everywhere except for prompt (but argparse expects it).

**Comment by blessedcoolant** [comment]
I made a 60 frame animation. The feature seems to be working fine. The zoom wasn't very obvious though. But it could have been my prompt that didn't really match what this feature is meant for.

But the issue with the folder name needs to be fixed if you want to do the frames to a named folder. This is a big problem on Windows with long folder names. The issue is being tackled here if you want to have a look. #189 

Don't want to approve this PR with the folder name issue. If you can't find a solution for that, maybe just bypass naming the folders and do a more generic folder for an output for the time being.

**Comment by veprogames** [comment]
Folder names are now in the format `video-{first word}-{timestamp}`

**Comment by blessedcoolant** [comment]
Thanks for fixing the folder issue. It seems fine now.

But for some reason I've started noticing some performance issues with the img2img in this implementation. At random intervals, the img2img decoding takes long pauses and then resumes. Wondering if it's just me.

Can someone else double check this. @lstein @tildebyte

**Comment by veprogames** [comment]
When I tested it, I didn't notice more pauses than using img2img through the dream interface, but I didn't test very intensively. I observed that more pauses happen when more file i/o is going on (using low force -> writing an image every few seconds)

**Comment by blessedcoolant** [comment]
> When I tested it, I didn't notice more pauses than using img2img through the dream interface, but I didn't test very intensively. I observed that more pauses happen when more file i/o is going on (using low force -> writing an image every few seconds)

There's a ton of other changes that are happening at the moment that might have been causing it. I'll take a look at your PR again today and do some testing.

If you're keen in the meanwhile, try to see if you can integrate this somehow into the main script using the -video prompt tag. It's a great feature and I think having this in the pipeline is a lot handier than as its own script.

**Comment by tildebyte** [comment]
@veprogames; As-is, this needs to be collapsed to a single commit, but don't waste time on it if you're refactoring to integrate it into the main script

**Comment by tildebyte** [comment]
@veprogames; What is "cv2 VideoWriter"? What package is it in, and how is it installed? I don't see anything matching on PyPi.

**Comment by veprogames** [comment]
> @veprogames; What is "cv2 VideoWriter"? What package is it in, and how is it installed? I don't see anything matching on PyPi.

VideoWriter is included in opencv-python which is a dependency in environment.yaml, so I used it to generate the video on the fly.

> If you're keen in the meanwhile, try to see if you can integrate this somehow into the main script using the -video prompt tag. It's a great feature and I think having this in the pipeline is a lot handier than as its own script.

I've looked at main.py but don't know where to start. If anyone wants to toy around with my code, go ahead.

**Comment by blessedcoolant** [comment]
> I've looked at main.py but don't know where to start. If anyone wants to toy around with my code, go ahead.

The files you need to look at are `dream.py` and `simplet2i.py`

**Comment by tildebyte** [comment]
> Create a .mp4 with cv2 VideoWriter (works at least on Windows)

I was pondering this, and wanted to look into it. AFAICT, opencv-python is cross-platform; the only potential issue that occurs to me would be if VideoWriter relies on system libs to create the final video (and if any of the sys libs are inconsistent across platforms, e.g. patent-encumbered and thus not distributed for some linuxes).

TL;DR - we need someone with a Linux box to test it :)

**Comment by smirnov-mv** [comment]
Tested this feature with parameters 
`"innsmouth, dramatic lighting, chiaroscuro, high detail, painted by greg rutkowski, painted by igor kieryluk, painted by bobby chiu, trending on artstation " -F 240 -f 0.8 -P 24` 
It's look cool but strange with some frames and with no smoothness.
https://user-images.githubusercontent.com/93412872/188032191-06fd3115-4637-4be8-a1e5-f227d85b2f80.mp4

I expected something like that https://www.reddit.com/r/MediaSynthesis/comments/x2yqlj/poisons_of_civilization_deforum_stable_diffusion/ https://colab.research.google.com/github/deforum/stable-diffusion/blob/main/Deforum_Stable_Diffusion.ipynb 

And i have several questions/ideas:
* how about parameter to zoom in and zoom out?
* how can I recreate video with different fps from exist frames? (for new video I must waiting new frames generation)
* how can I add new frames to exists to recreate video with more length? 
* how can I change width and height of frames and video?

And frame 0.png writing twice.

Seems like this feature should have 3 options for being more user friendly:
* generate frames (so I can delete bad frames)
* generate video from folder with parameters (fps, zoom_speed). In this case it should not be broken, when one or few frames was deleted, e.q.  1.png, 2.png, 5 png (3 and 4 I deleted when checked frames).
* generate frames and video.

May be name "video.mp4" in this case should be "video+fps+datetime.mp4"

**Comment by lstein** [comment]
Is this feature still working?  It would be great if there were a generic command line option similar to --grid that takes a series of images generated by -n and turns them into a movie. Then this would integrate well with the new variations feature without additional effort. What do you think?

BTW, we're asking PRs to go against the development branch to reduce new feature integration friction.

**Comment by veprogames** [comment]
The produced video doesn't give the expected results so far (see above comment; expected: https://www.reddit.com/r/MediaSynthesis/comments/x2yqlj/poisons_of_civilization_deforum_stable_diffusion/, got https://user-images.githubusercontent.com/93412872/188032191-06fd3115-4637-4be8-a1e5-f227d85b2f80.mp4)

If the smoothness is given, it would make sense to integrate this into dream CLI. Depends on motivation and capability.

**Comment by lstein** [comment]
Looks like this PR needs work. Converting into a draft.

**Comment by lstein** [comment]
Is any more work going to be done on this feature, or should I close it for now?

**Comment by veprogames** [comment]
I have added the remaining Tasks now, but I don't know how to solve the smoothness issue (low strength seemed to converge to single color images). I'll let you decide to close it or not, but I am out of ideas now about fixing the smoothness issue.

**Comment by smirnov-mv** [comment]
May be we can use this (https://github.com/deforum/stable-diffusion) realization or ideas?

**Review by lstein — CHANGES_REQUESTED** [review]
I'm happy to bring this in as a standalone script, but the code base has moved on and some minor changes are needed to make it work.
1. ldm.t2i has now changed to ldm.generate, and the object class name has changed too. While t2i is still there as a legacy mechanism, it is deprecated.
2. Create a new standalone .md file for the documentation and point to it from README.md. You could put the new doc into docs/features/DREAMVIDEO.md

Here's another idea that I'd like to bounce off you. We already have the `--grid` option in dream.py that collects all the images generated by a command and pastes them into a single gridded image. How about we have a `--video` option that takes all the outputs and turns them into a movie file? This would really work well with the -v option, which generates a bunch of minor variations, as well as with your idea of iteratively passing the output of one generation into the input of the next. Presumably `--video` could take all the options for framerate, etc, that you need.

**Comment by veprogames** [comment]
> 1. ldm.t2i has now changed to ldm.generate, and the object class name has changed too. While t2i is still there as a legacy mechanism, it is deprecated.
> 2. Create a new standalone .md file for the documentation and point to it from README.md. You could put the new doc into docs/features/DREAMVIDEO.md

I think the easiest way here would be to incoporate main into my branch first, by replacing everything except dream_video.py. (then there are also better inference times & less VRAM usage while testing). Is that done with a rebase?

I've got an idea about the jumpyness of the video: I currently use a random seed per frame. Does using the same seed throughout the entire video stabilize it? (compared to deforum stable diffusion)

**Comment by veprogames** [comment]
ok, I think I rebased 1.14 into my branch (my first rebase). Seems still to be a conflict with the README. Will try constant seed per video soon.

>  How about we have a --video option that takes all the outputs and turns them into a movie file?

I don't know if the dream CLI at one point could become too complex with many options and hard to maintain. I don't know if I am able to integrate it into the dream CLI myself. I'll adjust the tasks.

**Comment by veprogames** [comment]
(got something messed up, that's why it says --global as commit author, will be fixed on next commit)

**Comment by veprogames** [comment]
@lstein Since rebasing kinda messed this up (see all the conflicts), should I start fresh from dev and open a new PR? Or is there a way to "remove" the commits to limit changes to only the 3 files I've actually changed, and not 87

**Comment by veprogames** [comment]
Here's an example video: `python scripts/dream_video.py "Mountains by a Sea, painted by Greg Rutkowski" -F 60 -P 6 -f 0.8`

https://user-images.githubusercontent.com/75524847/190488262-28275b1d-3e0e-47ad-b58a-de2887f8c114.mp4


It seems to be smoother now.

> It's look cool but strange with some frames and with no smoothness.
https://user-images.githubusercontent.com/93412872/188032191-06fd3115-4637-4be8-a1e5-f227d85b2f80.mp4

could also work if the video was rendered with 1-2 FPS and then morphed together.

**Comment by lstein** [comment]
I'm afraid that too much time has passed and the number of conflicts with `development` has become problematic. I'm reluctantly closing this PR, but would be happy to accept a refreshed one. My apologies for how this worked out.

**Comment by veprogames** [comment]
Deforum Stable Diffusion seems to be much more advanced for making zoom videos, so I don't think it will be worth to bring simple zoom videos in here, Deforum will do a much better job at this.
